### DISCUSSION TITLE
unlock keyword

### DISCUSSION THREAD

**Discussion by mightywill** [dotnet/csharplang#7215] [Language Ideas]
Currently if you have some code that need to be completely thread-safe and you have time-consuming parts in it then you are enforced to wrap such code with try-finally statement:

```
	lock(_lock)
	{
		//  some thread-unsafe code,  e.g. prepare network request

		try
		{
			Monitor.Exit(_lock)

			// time-consuming code, e.g. requesting and waiting for a response
		}
		finally
		{
			Monitor.Enter(_lock)
		}

		//  some thread-unsafe code, e.g. processing a response
	}
```
but with `unlock` it would require much less code 

```
	lock(_lock)
	{
		//  some thread-unsafe code,  e.g. prepare network request

		unlock(_lock)
		{
			// time-consuming code, like requesting and waiting for a response
		}

		//  some thread-unsafe code
	}
```

**Comment by huoyaoyuan** [comment]
Then you will just have two `lock` scope, and the lock only provides thread safety within one scope. You still need to prove the invariants between two scopes.

In the other side, the best practice with network request is using async code. The response will resume one another thread and makes `lock` not working at all.

**Comment by Tvde1** [comment]
What's wrong with:
```cs
lock(_lock)
{
    //  some thread-unsafe code,  e.g. prepare network request
}

// time-consuming code, like requesting and waiting for a response
		
lock(_lock)
{
    //  some thread-unsafe code
}
```

**Comment by mightywill** [comment]
It's wrong when you deal with situations like this:
```
lock(_lock)
{
   if(...)
   {
      if(...)
      {
         if(...)
         {
            if(...)
            {
}

// time-consuming code, like requesting and waiting for a response
		
lock(_lock)
{
            }
         }
      }
   }
}
```

**Reply by huoyaoyuan** [reply]
Can you give more realistic example for what you are locking for? I don't see two separate locks common.
"Time consuming" isn't really related here. You are creating two atomic scopes, no matter what operations are done between them.
